uniffi_bindgen/src/crate_mode.rs

```diff
@@ -0,0 +1,244 @@
+    env::consts::{DLL_EXTENSION, DLL_PREFIX, DLL_SUFFIX},
+    fs,
+    process::{Command, Stdio},
+};
+use uniffi_meta::group_metadata;
+
+/// Generate foreign bindings
+///
+/// Returns the list of sources used to generate the bindings.  The main crate is first, followed
+/// by all other crates in no particular order.
+pub fn generate_bindings(
+    main_crate_name: &str,
+    target_languages: &[String],
+    out_dir: &Utf8Path,
+    try_format_code: bool,
+) -> Result<Vec<Source>> {
+    println!("generating bindings: {main_crate_name}");
+    let cargo_metadata = MetadataCommand::new()
+        .exec()
+        .context("error running cargo metadata")?;
+    let main_crate_package = find_package(&cargo_metadata, main_crate_name)?;
+    let library_path = find_cdylib_path(main_crate_name, &main_crate_package)?;
+    let mut sources = calc_sources(&cargo_metadata, &library_path)?;
+    for i in 0..sources.len() {
+        // Partition up the sources list because we're eventually going to call
+        // `update_from_dependency_configs()` which requires an exclusive reference to one source and
+        // shared references to all other sources.
+        let (sources_before, rest) = sources.split_at_mut(i);
+        let (source, sources_after) = rest.split_first_mut().unwrap();
+        let other_sources = sources_before.iter().chain(sources_after.iter());
+        // Calculate which configs come from dependent crates
+        let dependencies =
+            HashSet::<&str>::from_iter(source.package.dependencies.iter().map(|d| d.name.as_str()));
+        let config_map: HashMap<&str, &Config> = other_sources
+            .filter_map(|s| {
+                dependencies
+                    .contains(s.package.name.as_str())
+                    .then_some((s.crate_name.as_str(), &s.config))
+            })
+            .collect();
+        // We can finally call update_from_dependency_configs
+        source.config.update_from_dependency_configs(config_map);
+    }
+    fs::create_dir_all(out_dir)?;
+    // Move the main package to the front
+    let main_crate_index = sources
+        .iter()
+        .position(|s| s.package == main_crate_package)
+        .unwrap();
+    sources.swap(0, main_crate_index);
+    for source in sources.iter() {
+        for language in target_languages {
+            let language: bindings::TargetLanguage = language.as_str().try_into()?;
+            println!("{} {} {}", language, source.package.name, out_dir);
+            bindings::write_bindings(
+                &source.config.bindings,
+                &source.ci,
+                out_dir,
+                language,
+                try_format_code,
+            )?;
+        }
+    }
+
+    Ok(sources)
+}
+
+// A single source that we generate bindings for
+#[derive(Debug)]
+pub struct Source {
+    pub package: Package,
+    pub crate_name: String,
+    pub ci: ComponentInterface,
+    pub config: Config,
+}
+
+fn calc_sources(
```

Comment:
Not too happy with the name, as to me it's not really clear what we "calculate".
But I couldn't come up with a better name either.
`load_sources`? `parse_sources`? None really read much better, so maybe `calc_sources` is as good as any other.